File: app/services.py
import logging
import os
import time
from pathlib import Path
from typing import List

import requests
from dotenv import load_dotenv

from app.storage import IMAGES_DIR, load_groups, load_posts, save_posts
from app.vk_client import fetch_posts_for_groups


logger = logging.getLogger("vk_feed.services")


load_dotenv(override=False)


MAX_POSTS = int(os.getenv("MAX_POSTS", "500"))
MAX_IMAGES_PER_POST = 10


def _download_post_images(post: dict) -> List[str]:
    images: List[str] = []
    attachments = post.get("attachments") or []
    post_id = str(post.get("id"))

    for index, att in enumerate(attachments):
        if len(images) >= MAX_IMAGES_PER_POST:
            break

        if not isinstance(att, dict):
            continue

        if att.get("type") != "photo":
            continue

        url = att.get("url")
        if not url:
            continue

        filename = f"{post_id.replace('-', '_')}_{index}.jpg"
        local_path = IMAGES_DIR / filename

        if local_path.exists():
            images.append(filename)
            continue

        try:
            resp = requests.get(url, timeout=15)
            if resp.status_code == 200:
                IMAGES_DIR.mkdir(parents=True, exist_ok=True)
                with local_path.open("wb") as f:
                    f.write(resp.content)
                images.append(filename)
        except Exception as exc:
            logger.error("Failed to download image for post %s: %s", post_id, exc)

    return images


def update_posts_storage() -> int:
    group_objects = load_groups()
    if not group_objects:
        logger.info("No groups configured, skipping fetch")
        return 0

    existing_posts: List[dict] = load_posts()
    existing_ids = {str(post.get("id")) for post in existing_posts}

    fetched_posts = fetch_posts_for_groups(group_objects)

    now_ts = int(time.time())
    threshold = now_ts - 4 * 3600
    recent_fetched = [
        post
        for post in fetched_posts
        if isinstance(post.get("date"), (int, float)) and post["date"] >= threshold
    ]

    new_posts: List[dict] = []
    for post in recent_fetched:
        if str(post.get("id")) in existing_ids:
            continue
        post_images = _download_post_images(post)
        if post_images:
            post["postImages"] = post_images
        new_posts.append(post)

    if not new_posts:
        logger.info("No new posts fetched")
        return 0

    combined_posts = new_posts + existing_posts
    if MAX_POSTS > 0:
        combined_posts = combined_posts[:MAX_POSTS]

    save_posts(combined_posts)

    logger.info("Added %s new posts, total %s", len(new_posts), len(combined_posts))
    return len(new_posts)

File: src/lib/services.js
import fs from 'node:fs/promises'
import path from 'node:path'
import { appConfig } from './config.js'
import { fetchPostsForGroups } from './vk-client.js'
import { loadGroups, loadPosts, savePosts, loadSchedule } from './storage.js'

const maxImagesPerPost = 10

const resolvePostsImagesDir = () =>
  path.isAbsolute(appConfig.postsImagesDir)
    ? appConfig.postsImagesDir
    : path.resolve(appConfig.rootDir, appConfig.postsImagesDir)

const logImage = (message) => {
  process.stderr.write(`[news-feed images] ${message}\n`)
}

const downloadPostImages = async (post) => {
  const images = []
  const attachments = post?.attachments || []
  const postId = String(post?.id || '')
  const targetDir = resolvePostsImagesDir()

  for (let index = 0; index < attachments.length; index += 1) {
    if (images.length >= maxImagesPerPost) {
      break
    }

    const attachment = attachments[index]
    if (attachment?.type !== 'photo') {
      continue
    }
    if (!attachment?.url) {
      logImage(`post ${postId} attachment #${index}: type photo but no URL (VK payload missing sizes/orig_photo?)`)
      continue
    }

    const filename = `${postId.replaceAll('-', '_')}_${index}.jpg`
    const localPath = path.join(targetDir, filename)

    try {
      await fs.access(localPath)
      images.push(filename)
      continue
    } catch {
      // file missing, download below
    }

    try {
      const response = await fetch(attachment.url, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; NewsFeed/1.0)',
          Accept: 'image/avif,image/webp,image/apng,image/*,*/*;q=0.8',
        },
      })
      if (!response.ok) {
        logImage(`post ${postId} fetch ${response.status} ${response.statusText} — ${attachment.url.slice(0, 120)}`)
        continue
      }
      const arrayBuffer = await response.arrayBuffer()
      await fs.mkdir(targetDir, { recursive: true })
      await fs.writeFile(localPath, Buffer.from(arrayBuffer))
      images.push(filename)
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error)
      logImage(`post ${postId} save failed (${localPath}): ${msg}`)
    }
  }

  return images
}

export const updatePostsStorage = async () => {
  const groups = await loadGroups()
  if (!groups.length) {
    return 0
  }

  const existingPosts = await loadPosts()
  const existingIds = new Set(existingPosts.map((post) => String(post?.id)))
  const fetchedPosts = await fetchPostsForGroups(groups)
  const threshold = Math.floor(Date.now() / 1000) - (4 * 3600)
  const recentFetched = fetchedPosts.filter((post) => Number(post?.date) >= threshold)

  const newPosts = []
  let loggedImagesDir = false
  for (const post of recentFetched) {
    if (existingIds.has(String(post.id))) {
      continue
    }
    if (!loggedImagesDir) {
      process.stderr.write(`[news-feed] saving images under: ${resolvePostsImagesDir()} (cwd=${appConfig.rootDir})\n`)
      loggedImagesDir = true
    }
    const postImages = await downloadPostImages(post)
    if (postImages.length) {
      post.postImages = postImages
    }
    newPosts.push(post)
  }

  if (!newPosts.length) {
    if (recentFetched.length > 0) {
      process.stderr.write(
        '[news-feed] no new posts to save (all fetched wall items already in posts.json); images are only downloaded for newly added posts\n',
      )
    }
    return 0
  }

  const combined = [...newPosts, ...existingPosts].slice(0, appConfig.maxPosts)
  await savePosts(combined)
  return newPosts.length
}

export const runScheduledUpdate = async () => {
  const times = await loadSchedule()
  const now = new Date()
  const hh = String(now.getHours()).padStart(2, '0')
  const mm = String(now.getMinutes()).padStart(2, '0')
  const nowTime = `${hh}:${mm}`
  if (!times.includes(nowTime)) {
    return 0
  }
  return updatePostsStorage()
}
